File: tests/python/test_net_myrsi.py
"""Python binding tests for NET_MYRSI indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close
from rust_comparison import compare_with_rust


class TestNetMyrsi:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_net_myrsi_partial_params(self, test_data):
        """Test NET_MYRSI with partial parameters (None values) - mirrors check_net_myrsi_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.net_myrsi(close, 14, None)  
        assert len(result) == len(close)
    
    def test_net_myrsi_accuracy(self, test_data):
        """Test NET_MYRSI matches expected values from Rust tests - mirrors check_net_myrsi_accuracy"""
        close = test_data['close']
        
        
        period = 14
        
        result = ta_indicators.net_myrsi(close, period, None)
        
        assert len(result) == len(close)
        
        
        expected_last_five = [
            0.64835165,
            0.49450549,
            0.29670330,
            0.07692308,
            -0.07692308,
        ]
        
        
        assert_close(
            result[-5:], 
            expected_last_five,
            rtol=1e-7,  
            msg="NET_MYRSI last 5 values mismatch"
        )
        
        
        
    
    def test_net_myrsi_default_candles(self, test_data):
        """Test NET_MYRSI with default parameters - mirrors check_net_myrsi_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.net_myrsi(close, 14, None)
        assert len(result) == len(close)
    
    def test_net_myrsi_zero_period(self):
        """Test NET_MYRSI fails with zero period - mirrors check_net_myrsi_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.net_myrsi(input_data, 0, None)
    
    def test_net_myrsi_period_exceeds_length(self):
        """Test NET_MYRSI fails when period exceeds data length - mirrors check_net_myrsi_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.net_myrsi(data_small, 10, None)
    
    def test_net_myrsi_very_small_dataset(self):
        """Test NET_MYRSI with very small dataset - mirrors check_net_myrsi_very_small_dataset"""
        
        data_small = np.array([10.0, 20.0, 30.0, 15.0, 25.0])
        
        result = ta_indicators.net_myrsi(data_small, 3, None)
        assert len(result) == len(data_small)
        
        
        assert np.isnan(result[0])
        assert np.isnan(result[1])
        
        assert not np.all(np.isnan(result))
    
    def test_net_myrsi_empty_input(self):
        """Test NET_MYRSI with empty input"""
        input_data = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.net_myrsi(input_data, 14, None)
    
    def test_net_myrsi_all_nan(self):
        """Test NET_MYRSI with all NaN values"""
        input_data = np.array([np.nan] * 30)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.net_myrsi(input_data, 14, None)
    
    def test_net_myrsi_insufficient_data(self):
        """Test NET_MYRSI with insufficient data"""
        
        input_data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])  
        
        with pytest.raises(ValueError, match="(Invalid period|Not enough valid data)"):
            ta_indicators.net_myrsi(input_data, 10, None)  
    
    def test_net_myrsi_nan_handling(self):
        """Test NET_MYRSI handles NaN in middle of data - mirrors check_net_myrsi_nan_handling"""
        
        data = list(range(1, 11))  
        for i in range(20):
            data.append(data[-1] + 1.0)
        
        input_data = np.array(data, dtype=np.float64)
        period = 14
        
        
        data_with_nan = input_data.copy()
        data_with_nan[15] = np.nan
        
        result = ta_indicators.net_myrsi(data_with_nan, period, None)
        assert len(result) == len(data_with_nan)
        
        
        
        
        
        assert not np.all(np.isnan(result[:15])), "Should have valid values before NaN"
        assert not np.all(np.isnan(result[16:])), "Should have valid values after NaN"
        
        
        data_multi_nan = input_data.copy()
        data_multi_nan[10] = np.nan
        data_multi_nan[20] = np.nan
        
        
        result2 = ta_indicators.net_myrsi(data_multi_nan, period, None)
        assert len(result2) == len(data_multi_nan)
        assert isinstance(result2, np.ndarray)
    
    def test_net_myrsi_warmup_nans(self, test_data):
        """Test NET_MYRSI preserves warmup NaNs - mirrors check_net_myrsi_warmup_nans"""
        close = test_data['close']
        period = 14
        
        
        first_valid = 0
        for i, val in enumerate(close):
            if not np.isnan(val):
                first_valid = i
                break
        
        result = ta_indicators.net_myrsi(close, period, None)
        
        
        
        warmup = first_valid + period - 1
        
        
        for i in range(warmup):
            assert np.isnan(result[i]), f"Expected NaN at index {i} (warmup period)"
        
        
        
        actual_start = first_valid + period  
        if actual_start < len(result):
            assert not np.isnan(result[actual_start]), f"Expected valid value at index {actual_start} (first computed value)"
        
        
        if actual_start + 5 < len(result):
            for i in range(actual_start, actual_start + 5):
                assert not np.isnan(result[i]), f"Expected valid value at index {i} (after warmup)"
    
    def test_net_myrsi_stream(self, test_data):
        """Test NET_MYRSI streaming functionality"""
        close = test_data['close']
        period = 14
        
        
        stream = ta_indicators.NetMyrsiStream(period)
        
        
        stream_results = []
        for value in close:
            result = stream.update(value)
            stream_results.append(result if result is not None else np.nan)
        
        
        batch_results = ta_indicators.net_myrsi(close, period, None)
        
        
        
        stream_first_valid = next((i for i, v in enumerate(stream_results) if not np.isnan(v)), len(stream_results))
        batch_first_valid = next((i for i, v in enumerate(batch_results) if not np.isnan(v)), len(batch_results))
        
        
        first_valid = max(stream_first_valid, batch_first_valid)
        
        if first_valid < len(batch_results):
            
            
            assert_close(
                stream_results[first_valid:],
                batch_results[first_valid:],
                rtol=0,
                atol=1e-10,
                msg="Stream and batch results mismatch"
            )
    
    def test_net_myrsi_batch(self, test_data):
        """Test NET_MYRSI batch processing with metadata verification"""
        close = test_data['close']
        
        
        results = ta_indicators.net_myrsi_batch(close, (10, 20, 5), None)
        
        
        assert 'values' in results
        assert 'periods' in results
        
        
        assert results['values'].shape[0] == 3
        assert results['values'].shape[1] == len(close)
        assert len(results['periods']) == 3
        assert list(results['periods']) == [10, 15, 20]
        
        
        expected_periods = [10, 15, 20]
        for i, expected_period in enumerate(expected_periods):
            assert results['periods'][i] == expected_period, f"Period mismatch at index {i}"
        
        
        for i, period in enumerate(expected_periods):
            row = results['values'][i]
            
            warmup_end = period - 1
            assert np.all(np.isnan(row[:warmup_end])), f"Expected NaN in warmup for period {period}"
            
            
            if len(row) > warmup_end + 10:
                assert not np.all(np.isnan(row[warmup_end+10:])), f"Should have valid values for period {period}"
    
    def test_net_myrsi_batch_single_period(self, test_data):
        """Test NET_MYRSI batch with single period - mirrors ALMA batch tests"""
        close = test_data['close'][:100]  
        
        
        results = ta_indicators.net_myrsi_batch(close, (14, 14, 0), None)
        
        
        assert results['values'].shape[0] == 1
        assert results['values'].shape[1] == len(close)
        assert len(results['periods']) == 1
        assert results['periods'][0] == 14
        
        
        single_result = ta_indicators.net_myrsi(close, 14, None)
        
        
        
        assert len(results['values'][0]) == len(single_result)
        
        
        for i in range(13):  
            assert np.isnan(results['values'][0][i]) == np.isnan(single_result[i])
    
    def test_net_myrsi_kernel_consistency(self, test_data):
        """Test NET_MYRSI produces consistent results across different kernels"""
        close = test_data['close']
        period = 14
        
        
        
        kernels = [
            (None, "Auto"),      
            (1, "Scalar"),       
            (2, "SSE2"),         
        ]
        
        results = {}
        for kernel_value, kernel_name in kernels:
            try:
                result = ta_indicators.net_myrsi(close, period, kernel_value)
                results[kernel_name] = result
            except Exception as e:
                
                print(f"Kernel {kernel_name} not available: {e}")
                continue
        
        
        if len(results) > 1:
            kernel_names = list(results.keys())
            base_kernel = kernel_names[0]
            base_result = results[base_kernel]
            
            for kernel_name in kernel_names[1:]:
                
                for i in range(len(base_result)):
                    if not np.isnan(base_result[i]) and not np.isnan(results[kernel_name][i]):
                        assert_close(
                            base_result[i], 
                            results[kernel_name][i],
                            rtol=1e-12,
                            msg=f"Kernel {base_kernel} vs {kernel_name} mismatch at index {i}"
                        )
                    else:
                        
                        assert np.isnan(base_result[i]) == np.isnan(results[kernel_name][i]), \
                            f"NaN mismatch between {base_kernel} and {kernel_name} at index {i}"
    
    def test_net_myrsi_numerical_precision(self, test_data):
        """Test NET_MYRSI numerical precision and edge cases"""
        
        extreme_data = np.array([1e-10, 1e10, 1e-10, 1e10] * 10, dtype=np.float64)
        result = ta_indicators.net_myrsi(extreme_data, 5, None)
        assert len(result) == len(extreme_data)
        
        assert not np.any(np.isinf(result[~np.isnan(result)])), "Should not produce infinity"
        
        
        small_diff_data = np.array([100.0 + i * 1e-10 for i in range(50)], dtype=np.float64)
        result = ta_indicators.net_myrsi(small_diff_data, 10, None)
        assert len(result) == len(small_diff_data)
        
        valid_values = result[~np.isnan(result)]
        if len(valid_values) > 0:
            
            
            assert not np.any(np.isinf(valid_values)), "Should not produce infinity"
            assert not np.any(np.isnan(valid_values)), "Valid values should not be NaN"
        
        
        constant_data = np.full(30, 100.0, dtype=np.float64)
        result = ta_indicators.net_myrsi(constant_data, 10, None)
        assert len(result) == len(constant_data)
        
        
        valid_values = result[~np.isnan(result)]
        if len(valid_values) > 0:
            assert not np.any(np.isinf(valid_values)), "Should not produce infinity with constant values"
            
            assert np.all(valid_values >= -1.0), "NET_MYRSI should be >= -1"
            assert np.all(valid_values <= 1.0), "NET_MYRSI should be <= 1"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/wasm/test_net_myrsi.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('NET_MYRSI partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.net_myrsi_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('NET_MYRSI accuracy', () => {

    const close = new Float64Array(testData.close);


    const period = 14;

    const result = wasm.net_myrsi_js(close, period);

    assert.strictEqual(result.length, close.length);


    const expected_last_five = [
        0.64835165,
        0.49450549,
        0.29670330,
        0.07692308,
        -0.07692308,
    ];


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected_last_five,
        1e-7,
        "NET_MYRSI last 5 values mismatch"
    );
});

test('NET_MYRSI default params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.net_myrsi_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('NET_MYRSI zero period', () => {

    const input_data = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.net_myrsi_js(input_data, 0);
    }, /Invalid period/, 'Should throw error for zero period');
});

test('NET_MYRSI period exceeds length', () => {

    const data_small = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.net_myrsi_js(data_small, 10);
    }, /Invalid period/, 'Should throw error when period exceeds data length');
});

test('NET_MYRSI very small dataset', () => {


    const data_small = new Float64Array([10.0, 20.0, 30.0, 15.0, 25.0]);

    const result = wasm.net_myrsi_js(data_small, 3);
    assert.strictEqual(result.length, data_small.length);


    assert(isNaN(result[0]), 'First value should be NaN');
    assert(isNaN(result[1]), 'Second value should be NaN');

    assert(!result.every(isNaN), 'Should have some valid values');
});

test('NET_MYRSI empty input', () => {

    const input_data = new Float64Array([]);

    assert.throws(() => {
        wasm.net_myrsi_js(input_data, 14);
    }, /Input data slice is empty/, 'Should throw error for empty input');
});

test('NET_MYRSI all NaN', () => {

    const input_data = new Float64Array(30);
    input_data.fill(NaN);

    assert.throws(() => {
        wasm.net_myrsi_js(input_data, 14);
    }, /All values are NaN/, 'Should throw error for all NaN input');
});

test('NET_MYRSI insufficient data', () => {


    const input_data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);

    assert.throws(() => {
        wasm.net_myrsi_js(input_data, 10);
    }, /(Invalid period|Not enough valid data)/, 'Should throw error for insufficient data');
});

test('NET_MYRSI nan handling', () => {


    const data = new Float64Array(30);
    for (let i = 0; i < 10; i++) {
        data[i] = i + 1.0;
    }
    for (let i = 10; i < 30; i++) {
        data[i] = data[i - 1] + 1.0;
    }

    const period = 14;


    const dataWithNaN = new Float64Array(data);
    dataWithNaN[15] = NaN;

    const result = wasm.net_myrsi_js(dataWithNaN, period);
    assert.strictEqual(result.length, dataWithNaN.length);





    let hasValidBefore = false;
    for (let i = 0; i < 15 && i < result.length; i++) {
        if (!isNaN(result[i])) {
            hasValidBefore = true;
            break;
        }
    }
    assert(hasValidBefore, 'Should have valid values before NaN');

    let hasValidAfter = false;
    for (let i = 16; i < result.length; i++) {
        if (!isNaN(result[i])) {
            hasValidAfter = true;
            break;
        }
    }
    assert(hasValidAfter, 'Should have valid values after NaN');


    const dataMultiNaN = new Float64Array(data);
    dataMultiNaN[10] = NaN;
    dataMultiNaN[20] = NaN;


    const result2 = wasm.net_myrsi_js(dataMultiNaN, period);
    assert.strictEqual(result2.length, dataMultiNaN.length);
    assert(Array.isArray(result2) || result2 instanceof Float64Array, 'Should return array');
});

test('NET_MYRSI warmup nans', () => {

    const close = new Float64Array(testData.close);
    const period = 14;


    let first_valid = 0;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            first_valid = i;
            break;
        }
    }

    const result = wasm.net_myrsi_js(close, period);



    const warmup = first_valid + period - 1;


    for (let i = 0; i < warmup; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} (warmup period)`);
    }



    const actual_start = first_valid + period;
    if (actual_start < result.length) {
        assert(!isNaN(result[actual_start]),
               `Expected valid value at index ${actual_start} (first computed value)`);
    }


    if (actual_start + 5 < result.length) {
        for (let i = actual_start; i < actual_start + 5; i++) {
            assert(!isNaN(result[i]), `Expected valid value at index ${i} (after warmup)`);
        }
    }
});

test('NET_MYRSI with NaN prefix', () => {

    const close = new Float64Array(testData.close);


    close[0] = NaN;
    close[1] = NaN;
    close[2] = NaN;

    const result = wasm.net_myrsi_js(close, 14);
    assert.strictEqual(result.length, close.length);


    assert(isNaN(result[0]), 'First value should be NaN');
    assert(isNaN(result[1]), 'Second value should be NaN');
    assert(isNaN(result[2]), 'Third value should be NaN');
});

test('NET_MYRSI memory allocation functions', () => {

    const len = 100;


    const ptr = wasm.net_myrsi_alloc(len);
    assert(ptr !== 0, 'Pointer should not be null');


    wasm.net_myrsi_free(ptr, len);

});

test('NET_MYRSI consistency check', () => {

    const close = new Float64Array(testData.close);
    const period = 14;

    const result1 = wasm.net_myrsi_js(close, period);
    const result2 = wasm.net_myrsi_js(close, period);

    assertArrayClose(
        result1,
        result2,
        1e-10,
        "Results should be identical for same input"
    );
});

test('NET_MYRSI different periods', () => {

    const close = new Float64Array(testData.close);

    const result10 = wasm.net_myrsi_js(close, 10);
    const result14 = wasm.net_myrsi_js(close, 14);
    const result20 = wasm.net_myrsi_js(close, 20);

    assert.strictEqual(result10.length, close.length);
    assert.strictEqual(result14.length, close.length);
    assert.strictEqual(result20.length, close.length);


    const last10 = result10[result10.length - 1];
    const last14 = result14[result14.length - 1];
    const last20 = result20[result20.length - 1];

    assert(Math.abs(last10 - last14) > 1e-10, 'Results should differ for different periods');
    assert(Math.abs(last14 - last20) > 1e-10, 'Results should differ for different periods');
});





